### DISCUSSION TITLE
MOZILLA_PKIX_ERROR_REQUIRED_TLS_FEATURE_MISSING

### DISCUSSION THREAD

**Discussion by Joly0** [ZoeyVid/NPMplus#627] [Q&A] [answered]
Hey, i keep getting the above error when i try to connect to any website behind npmplus through firefox.

![grafik](https://github.com/ZoeyVid/NPMplus/assets/13993216/1b29c162-0a95-42d1-b301-7eb2b6293be7)

Here is the docker logs:
```
nginx: [warn] "ssl_stapling" ignored, issuer certificate not found for certificate "/data/tls/dummycert.pem"
nginx: [warn] "ssl_stapling" ignored, issuer certificate not found for certificate "/data/tls/dummycert.pem"
nginx: [warn] "ssl_stapling" ignored, issuer certificate not found for certificate "/data/tls/dummycert.pem"
2024/01/23 18:40:25 [warn] 754#754: "ssl_stapling" ignored, issuer certificate not found for certificate "/data/tls/dummycert.pem"
2024/01/23 18:40:25 [warn] 754#754: "ssl_stapling" ignored, issuer certificate not found for certificate "/data/tls/dummycert.pem"
2024/01/23 18:40:25 [warn] 754#754: "ssl_stapling" ignored, issuer certificate not found for certificate "/data/tls/dummycert.pem"
```

**Comment by Zoey2936** [comment]
do you run a local instance?

**Comment by Joly0** [comment]
Found the issue. Its right there in the text. There are some leftover "dummycert.pem" files from npm. Not sure, why they are there, but i deleted them and now its working

**Comment by Zoey2936** [comment]
> do you run a local instance?

**Comment by Joly0** [comment]
Ok, i have to reopen this, as it is happening again. The dummycert.pem and dummykey.pem file appeared again. But deleting them didnt help this time. Additionaly, since switching to npmplus i cant access my jellyfin instance anymore. I always receive error code 400

**Comment by Joly0** [comment]
> do you run a local instance?

What exactly do you mean by "local instance"?

**Comment by Zoey2936** [comment]
do you access it using an private ip / domain that rewrites to a private ip

**Comment by Joly0** [comment]
Still not 100% sure what you mean, but i´ll try to explain. So i have npmplus and jellyfin on my homeserver. I have my pihole configured for local dns so when i access jellyfin.mydomain.de it returns the home server ip. But i already tested it without that route over pihole

**Comment by Zoey2936** [comment, marked as answer]
see README.md:
> Note: Internal/LAN Instance? Please disable must-staple in /opt/npm/tls/certbot/config.ini.

**Comment by Joly0** [comment]
> see README.md:
> 
> > Note: Internal/LAN Instance? Please disable must-staple in /opt/npm/tls/certbot/config.ini.

Will try that

**Comment by Joly0** [comment]
Ok, looks like that helped. Also the error 400 was caused by adding this proxy header "proxy_set_header Host $host;"
Though i am not sure, why this now only happens with npmplus and didnt with npm, but nontheless, its working now

**Comment by Zoey2936** [comment]
can I close?

**Comment by Joly0** [comment]
Btw, this error is happening again, although i set "must-staple = false" in config.ini

**Comment by Zoey2936** [comment]
did you just set it to false or did you also recreated your certs after that? (not renewed but recreated)

**Comment by Joly0** [comment]
I have one certbot certificate, which i renewed after setting it to false. But it doesnt say anywere in the readme to reacreate them, does it? The readme only mention custom certificates `if you use custom certificates, you need to upload the CA/Intermediate Certificate (file name: chain.pem) in the /opt/npm/tls/custom/npm-[certificate-id] folder`

**Comment by Zoey2936** [comment]
I've added it to the readme

**Comment by Joly0** [comment]
Ok, but now i keep getting this, when i try to set this new certificate for my proxy hosts and save it:
![grafik](https://github.com/ZoeyVid/NPMplus/assets/13993216/e8d4fb92-a5f4-4396-bf6e-b0230361ed9d)

```
nginx: [emerg] cannot load certificate "/data/tls/certbot/live/npm-11/fullchain.pem": BIO_new_file() failed (SSL: error:80000002:system library::No such file or directory:calling fopen(/data/tls/certbot/live/npm-11/fullchain.pem, r) error:10000080:BIO routines::no such file)
nginx: configuration file /usr/local/nginx/conf/nginx.conf test failed
```

**Comment by Zoey2936** [comment]
"No such file or directory"

**Comment by Joly0** [comment]
Restarted Container, and its not working anymore. In docker logs its stuck here:

```
----------------------------------------------------------
Let's Encrypt certificates cleanup - 2024/02/04 14:15:57
----------------------------------------------------------
Keeping /data/tls/certbot/archive/npm-10/cert4.pem.
Keeping /data/tls/certbot/archive/npm-10/chain4.pem.
Keeping /data/tls/certbot/archive/npm-10/fullchain4.pem.
Keeping /data/tls/certbot/archive/npm-10/privkey4.pem.
Keeping /data/tls/certbot/archive/npm-13/cert1.pem.
Keeping /data/tls/certbot/archive/npm-13/chain1.pem.
Keeping /data/tls/certbot/archive/npm-13/fullchain1.pem.
Keeping /data/tls/certbot/archive/npm-13/privkey1.pem.
8 file(s) kept.
0 file(s) deleted.
no DEFAULT_CERT_ID set, using dummycerts.
Working on file: /data/nginx/custom/events.conf
Working on file: /data/nginx/custom/http.conf
Working on file: /data/nginx/custom/http_top.conf
Working on file: /data/nginx/custom/root.conf
Working on file: /data/nginx/custom/server_dead.conf
Working on file: /data/nginx/custom/server_proxy.conf
Working on file: /data/nginx/custom/server_redirect.conf
Working on file: /data/nginx/custom/server_stream.conf
Working on file: /data/nginx/custom/server_stream_tcp.conf
Working on file: /data/nginx/custom/server_stream_udp.conf
Working on file: /data/nginx/custom/stream.conf
Working on file: /data/nginx/custom/stream_top.conf
Working on file: /data/nginx/default.conf
Working on file: /data/nginx/ip_ranges.conf
Working on file: /data/nginx/proxy_host/1.conf
Working on file: /data/nginx/proxy_host/11.conf
Working on file: /data/nginx/proxy_host/12.conf
Working on file: /data/nginx/proxy_host/2.conf
Working on file: /data/nginx/proxy_host/3.conf
Working on file: /data/nginx/proxy_host/5.conf
Working on file: /data/nginx/proxy_host/8.conf
Success.
usermod: no changes

-------------------------------------
 _ _  ___  __ __       _
| \ || . \|  \  \ ___ | | _ _  ___
|   ||  _/|     || . \| || | |[_-[
|_\_||_|  |_|_|_||  _/|_| \__|/__/
                 |_|
-------------------------------------
Version:  2.11.1+a779df8
Date:     Sun Feb  4 14:15:57 CET 2024
User:     npm
PUID:     99
User ID:  99
PGID:     100
Group ID: 100
-------------------------------------


```

And the admin page isnt loading anymore

**Comment by Joly0** [comment]
> "No such file or directory"

Yes, i saw that aswell and the file it is searching isnt there. I just deleted the old certificate and added a new one through certbot

**Comment by Zoey2936** [comment]
you wrote your own error
if you delete the cert, then nginx can't find it....

**Comment by Zoey2936** [comment]
run `rm -v /opt/npmplus/nginx/proxy_host/*`, restart npmplus, open npmplus on port 81, then create your cert, then update all your hosts with the new cert

**Reply by Joly0** [reply]
```
/app # rm -v /opt/nginx/proxy_host/*
rm: cannot remove '/opt/nginx/proxy_host/*': No such file or directory
/app # 
```

**Reply by Zoey2936** [reply]
I've fixed the command, it was meant to be run on the host with the default paths of the compose.yaml. Inside the container it needs to be `rm -v /data/nginx/proxy_host/*`

**Reply by Joly0** [reply]
Ok, worked

**Comment by Joly0** [comment]
> you wrote your own error if you delete the cert, then nginx can't find it....

Tbf, you wrote "recreate" and recreate for me means "delete old, create new". Otherwise you should have said "create a second one and leave the old one there", but yes, i see my fault there

**Comment by Paillat-dev** [comment]
Is it normal that I have to set `must-staple = false` even tough i am not in any custom / enterprise local instance with any custom cert?

**Reply by Zoey2936** [reply]
how do you exactly access your proxy hosts?

**Reply by Paillat-dev** [reply]
I'm sorry but I don't get what you mean.

**Reply by Zoey2936** [reply]
what are your dns settings?

**Reply by Paillat-dev** [reply]
```
   Serveurs DNS. . .  . . . . . . . . . . : 130.223.8.20
                                            130.223.4.5
```
I just access them trough the normal internet, as I said no intranet, nothing.

**Reply by Zoey2936** [reply]
then it should work... does your dns work correctly? can you send me your docker logs?

**Reply by Paillat-dev** [reply]
I will have to enable again the option and test again, I will asap.

**Comment by vrelk** [comment]
For those that are getting the internal error, that looks like something that @Zoey2936 needs to add a fix for. If you delete the cert before either removing or changing on from any hosts that are using it, you will get that error. I was able to work around this so I didn't have to delete all of the hosts as suggested in another comment, I will say what I did below.

For me, the location of these files on the host would be `/opt/npm/tls/certbot/live`, but it could be something like `/var/lib/docker/volumes/npm_data/_data/tls/certbot/live` if you used a docker volume. So either open a shell to the docker container, or do it on the host as root, but cd to the `live` folder.
1. Look at the docker logs to see what cert is missing. ex. `cannot load certificate "/data/tls/certbot/live/npm-3/fullchain.pem"`
2. do `ls` to see what other certs are available. If you don't have one, go to the web UI and make a new one now that you have disabled must-staple.
3. Copy another cert folder to the path of the one that it is trying to load. ex. `cp -r npm-4 npm-3`.
4. Go into the UI and update the cert of any site that needs the cert to the new one (well, any site that still references the missing cert. The UI says none, but the config file still shows the old cert.)
5. Go and delete the destination folder for the cert you copied in step 3. (This would be `npm-3` NOT `npm-4`)
6. Done

**Reply by Zoey2936** [reply]
1. yes, it is known, just create the new cert before removing the old cert
2. this will be fixed with the php rewrite or if someone provides a pr/it gets fixed in upstream
3. this is a different issue then this thread is about
4. the issue of this thread was fixed

**Comment by Joly0** [comment]
Btw, for some reason my initial issue is happening again. From my internal network i cant access my websites using their domain names. I again get the error "MOZILLA_PKIX_ERROR_SELF_SIGNED_CERT". must_staple is still configured to false (i see this is now the default) and i already tried recreating my ssl certificate through certbot, but it aint working.

**Reply by Joly0** [reply]
Edge shows this error "ERR_ECH_FALLBACK_CERTIFICATE_INVALID" and chrome shows this "ERR_QUIC_PROTOCOL_ERROR" followed by the same error as edge

**Reply by Zoey2936** [reply]
MOZILLA_PKIX_ERROR_SELF_SIGNED_CERT means SELF_SIGNED_CERT, do you uploaded a certiticate by hand? or does NPMplus maybe send you the default request using the dumycerts?

**Reply by Zoey2936** [reply]
can you maybe force renwe your cert?

**Reply by Joly0** [reply]
If you mean clicking "Renew now" on the tls certificates tab for the old and new certificate, then i have done that already.
Also i did not upload anything by hand. I had and have always been adding my certificates to npm and npmplus through certbot using cloudflare and the cloudflare api key that i have

**Reply by Joly0** [reply]
Ok, i figured it has something to do with cloudflare tunnels and tls1.3.
When i use cloudflare tunnel and have enabled tls1.3 at the same time, it does not work and give the error, if i dont use cloudflare tunnel disabled the tls1.3 settings is irrelevant, and if i have cloudflare tunnel enabled and tls1.3 disabled, it is working without errors aswell.

**Reply by Zoey2936** [reply]
So no issue on my side, right?

**Reply by Joly0** [reply]
It looks like no, but i am not sure why this is happening now when tls1.3 is enabled. I would like to keep it enabled

**Comment by Tumtum2814** [comment]
so i fixed my cert stapling issue, created new ones and set my hosts to use the new certs. 
It seems, somewhat randomly, NPM reloads itself. which then causes pages to display the `MOZILLA_PKIX_ERROR_REQUIRED_TLS_FEATURE_MISSING` error temporarily , when i refresh the error goes away.
Any ideas what might be happening?

**Reply by Zoey2936** [reply]
the way stapling works will change with the next release
